Source/WTF/wtf/text/CStringView.h

```diff
@@ -0,0 +1,113 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <span>
+#include <string>
+#include <type_traits>
+#include <wtf/Compiler.h>
+#include <wtf/ForbidHeapAllocation.h>
+#include <wtf/Forward.h>
+#include <wtf/HashFunctions.h>
+#include <wtf/StdLibExtras.h>
+#include <wtf/text/ASCIILiteral.h>
+#include <wtf/text/SuperFastHash.h>
+
+namespace WTF {
+
+class PrintStream;
+class String;
+
+class CStringView final {
+    WTF_FORBID_HEAP_ALLOCATION;
+public:
+    static CStringView unsafeFromUTF8(const char* string)
+    {
+        if (!string)
+            return CStringView();
+        return CStringView(unsafeMakeSpan(byteCast<char8_t>(string), std::char_traits<char>::length(string) + 1));
+    }
+
+    WTF_EXPORT_PRIVATE void dump(PrintStream& out) const;
+
+    CStringView() = default;
+    constexpr CStringView(std::nullptr_t)
+        : CStringView()
+    { }
+    CStringView(ASCIILiteral literal LIFETIME_BOUND)
+    {
+        if (!literal.length())
+            return;
+        m_charactersWithNullTerminator = byteCast<char8_t>(literal.spanIncludingNullTerminator().data());
+        m_lengthWithNullTerminator = literal.spanIncludingNullTerminator().size();
+    }
+
+    unsigned hash() const;
+    bool isNull() const { return !m_lengthWithNullTerminator; }
+
+    const char* rawCharacters() const LIFETIME_BOUND { return reinterpret_cast<const char*>(m_charactersWithNullTerminator); }
+    size_t length() const { return m_lengthWithNullTerminator > 0 ? m_lengthWithNullTerminator - 1 : 0; }
+    std::span<const char8_t> span8() const LIFETIME_BOUND { return unsafeMakeSpan(m_charactersWithNullTerminator, length()); }
+    size_t isEmpty() const { return m_lengthWithNullTerminator <= 1; }
+    WTF_EXPORT_PRIVATE String toString() const;
+
+    explicit operator bool() const { return !isEmpty(); }
+    bool operator!() const { return isEmpty(); }
+
+private:
+    explicit CStringView(std::span<const char8_t> spanWithNullTerminator LIFETIME_BOUND)
+        : m_charactersWithNullTerminator(spanWithNullTerminator.data())
+        , m_lengthWithNullTerminator(spanWithNullTerminator.size())
+    {
+    }
+
+    const char8_t* m_charactersWithNullTerminator { nullptr };
+    size_t m_lengthWithNullTerminator { 0 };
```

Comment:
For consistency, clarity, and efficiency, let's use the strategy we use in `CStringBuffer`, and have `m_length` hold the number of characters (which can be zero).

Reply:
I used the same strategy as in ASCIILiteral, which is another class that keeps the null terminated characters and accounts for that in the name (and meaning, obviously).

Reply:
I did the changes, to distinguish null and empty strings I chose to use -1 but we could switch to `std::optional<size_t>`. Anyway, my preference was to keep the null terminator accounted at the internal length as it is done for ASCIILiteral.

Reply:
I think I addressed this. I think the length is ok with the current implementation from the logical PoV.